src/toil/fileStores/cachingFileStore.py

```diff
@@ -97,1172 +113,1457 @@ class CachingFileStore(AbstractFileStore):
             self.jobStore.readFile(emptyID, cachedFile, symlink=False)
 
             # Check the link count
-            self.nlinkThreshold = os.stat(cachedFile).st_nlink
-
-            # Only 1 or 2 is allowed.
-            assert(self.nlinkThreshold == 1 or self.nlinkThreshold == 2)
+            if os.stat(cachedFile).st_nlink == 2:
+                # Caching must be free
+                free = 1
+            else:
+                # If we only have one link, caching costs disk.
+                free = 0
 
             # Clean up
             os.unlink(cachedFile)
+            os.rmdir(destDir)
             self.jobStore.deleteFile(emptyID)
         else:
-            # Unless we are on the file job store, we need to have a link count threshold of 1.
-            # TODO: This relies on something the job store interface doesn't actually guarantee!
-            self.nlinkThreshold = 1
+            # Caching is only ever free with the file job store
+            free = 0
+
+        # Save to the database if we're the first to work this out
+        self.cur.execute('INSERT OR IGNORE INTO properties VALUES (?, ?)', ('freeCaching', free))
+        self.con.commit()
+
+        # Return true if we said caching was free
+        return free == 1
+
+
+    # Internal caching logic
+
+    def _getNewCachingPath(self, fileStoreID):
+        """
+        Get a path at which the given file ID can be cached.
 
-    def _accountForNlinkEquals2(self, localFilePath):
+        Will be unique for every call. 
+
+        The file will not be created if it does not exist.
+        """
+
+        # Hash the file ID
+        hasher = hashlib.sha1()
+        hasher.update(fileStoreID.encode('utf-8'))
+
+        # Get a unique temp file name, including the file ID's hash to make
+        # sure we can never collide even though we are going to remove the
+        # file.
+        # TODO: use a de-slashed version of the ID instead?
+        handle, path = tempfile.mkstemp(dir=self.localCacheDir, suffix=hasher.hexdigest())
+        os.close(handle)
+        os.unlink(path)
+
+        return path
+    
+    def _stealWorkFromTheDead(self):
+        """
+        Take ownership of any files we can see whose owners have died.
+        
+        We don't actually process them here. We take action based on the states of files we own later.
         """
-        This is a utility function that accounts for the fact that if nlinkThreshold == 2, the
-        size of the file is accounted for by the file store copy of the file and thus the file
-        size shouldn't be added to the cached file sizes.
-        :param str localFilePath: Path to the local file that was linked to the file store copy.
+        
+        pid = os.getpid()
+        
+        # Get a list of all file owner processes on this node.
+        # Exclude NULL because it comes out as 0 and we can't look for PID 0.
+        owners = []
+        for row in self.cur.execute('SELECT DISTINCT owner FROM
```

Comment:
We check userPath here since this is the code that handles being given an explicit destination by the user (in which case we need to validate it and make it an absolute path).

`readGlobalFile` needs to write the retrieved file somewhere for the job to get at it (as opposed to `readGlobalFileStream` which just streams it). Here we're refusing to let the user overwrite an existing file by downloading a new file on top of it.